### PR TITLE
feat(SMT): reverse mutations generation, mutations serialization

### PR DESCRIPTION
**Pull request by polydez**
Related to https://github.com/0xPolygonMiden/miden-node/issues/527

While refactoring account inclusion proofs in-memory storage, I realized that it's very hard to keep persistent state for additional "initial" structure, and it would be easier to keep only reverse mutation sets and the latest state. Thus, we need to store only reverse mutations set, but latest state we can still reconstruct from the database.

In this PR I implemented reverse mutations set generation on mutations applying (using a new method `apply_mutations_with_reversion()`. Also implemented serialization of mutations set and added benchmarks for mutations calculation/applying.

### CODE REVIEW

File: src/merkle/smt/mod.rs [hunk 1]
```diff
@@ -287,20 +288,41 @@ pub(crate) trait SparseMerkleTree<const DEPTH: u8> {
...
-                Removal => self.remove_inner_node(index),
-                Addition(node) => self.insert_inner_node(index, node),
+                Removal => {
+                    if let Some(node) = self.remove_inner_node(index) {
+                        reverse_mutations.insert(index, Addition(node));
+                    }
+                },
+                Addition(node) => {
+                    if let Some(old_node) = self.insert_inner_node(index, node) {
+                        reverse_mutations.insert(index, Addition(old_node));
```
Comment (polydez) [inline src/merkle/smt/mod.rs:301]: Here we rely on the computation of mutations set, that it didn't generate useless update where new value is the same as previous (actually it does generate unnecessary changes in mutations set). Otherwise we would need to clone `node` and compare it with `old_node`.

File: src/merkle/smt/mod.rs [hunk 2]
```diff
@@ -287,20 +288,41 @@ pub(crate) trait SparseMerkleTree<const DEPTH: u8> {
...
+                    }
+                },
             }
         }
 
+        let mut reverse_pairs = BTreeMap::new();
         for (key, value) in new_pairs {
-            self.insert_value(key, value);
+            if let Some(old_value) = self.insert_value(key.clone(), value) {
+                reverse_pairs.insert(key, old_value);
```
Comment (polydez) [inline src/merkle/smt/mod.rs:312]: Here we also rely on the computation of mutations set, that it didn't generate useless update where new value is the same as previous (actually it does generate unnecessary changes in mutations set). Otherwise we would need to clone `value` and compare it with `old_value`.

File: CHANGELOG.md [hunk 3]
```diff
@@ -1,3 +1,7 @@
+# TBD
```
Comment (bobbinth) [inline CHANGELOG.md:1]: nit: let's change this to `## Unreleased`.

File: src/merkle/smt/mod.rs (hunk 1 above)
Comment (bobbinth) [reply src/merkle/smt/mod.rs:301]: > actually it does generate unnecessary changes in mutations set

Does this mean we could improve `compute_mutations()` method?

File: src/merkle/smt/mod.rs [hunk 4]
```diff
@@ -256,7 +257,7 @@ pub(crate) trait SparseMerkleTree<const DEPTH: u8> {
     }
 
     /// Apply the prospective mutations computed with [`SparseMerkleTree::compute_mutations()`] to
-    /// this tree.
+    /// this tree. Return reverse mutation set.
```
Comment (bobbinth) [inline src/merkle/smt/mod.rs:260]: nit: I would make this a single sentence:
```rust
/// Applies the prospective mutations computed with [`SparseMerkleTree::compute_mutations()`] to
/// this tree and returns the reverse mutation set.
```

I would also add a short paragraph (1 - 2 sentences) to explain what "reverse mutation set" is.

**Review by bobbinth — COMMENTED** [review]
Looks good! Thank you! Not an in-depth review yet, but I left a few small comments inline.

> The only concern I have is that now we produce reverse mutations set on each applying of mutations set. For the node it's fine, because we always need them after changing of the latest accounts SMT. But if someone else uses our SMT, it might be unnecessary to calculate such reversion on each mutations set applying.

Yeah - I was thinking about this as well. Could you run benchmarks to see how this affects things - especially for relatively large trees? (we may already have benchmarks for this).

If the effect is significant, I wonder if there is some clean interface to let the user apply mutations with or without generating the reverse set. (but let's run the benchmarks first).

File: src/merkle/smt/mod.rs (hunk 1 above)
Comment (polydez) [reply src/merkle/smt/mod.rs:301]: Yes, you're right, we could improve it in order to filter out unnecessary mutations (change value from `A` to `A`).

**Comment by polydez** [comment]
> If the effect is significant, I wonder if there is some clean interface to let the user apply mutations with or without generating the reverse set.

One of solutions would be introducing new `feature` for that behavior.

**Comment by bobbinth** [comment]
> One of solutions would be introducing new `feature` for that behavior.

As discussed offline, having 2 separate methods would be a more flexible solution (because there may be situations where the same codebase may want to use different versions of the method). Maybe, under the hood, both methods would use a single method that could look something like:
```rust
fn apply_mutations(
    &mut self,
    mutations: MutationSet<DEPTH, Self::Key, Self::Value>,
    return_inverse_mutations: bool,
) -> Result<MutationSet<DEPTH, Self::Key, Self::Value>, MerkleError>
```

**Comment by polydez** [comment]
> Could you run benchmarks to see how this affects things - especially for relatively large trees? (we may already have benchmarks for this).
> 
> If the effect is significant, I wonder if there is some clean interface to let the user apply mutations with or without generating the reverse set. (but let's run the benchmarks first).

I implemented [benchmark](https://github.com/0xPolygonMiden/crypto/pull/355/commits/a5595b9644c7bf3d24d2d7c061cfac00524531a0) for `apply_mutations` and here are the results on MacBook Pro with Apple M3 Max (unnecessary lines were removed):

---

**With** generation of reverse mutations set:

apply_mutations/SimpleSmt: apply_mutations/1000
                        time:   [30.044 ms 30.826 ms 31.697 ms]

apply_mutations/SimpleSmt: apply_mutations/100000
                        time:   [3.7554 s 3.8398 s 3.9335 s]

**Without** generation of reverse mutations set:

apply_mutations/SimpleSmt: apply_mutations/1000
                        time:   [23.827 ms 24.089 ms 24.289 ms]
                        change: [-5.3989% -3.4174% -1.7821%] (p = 0.00 < 0.05)
                        Performance has improved.

apply_mutations/SimpleSmt: apply_mutations/100000
                        time:   [3.2683 s 3.4088 s 3.5618 s]
                        change: [-15.168% -11.224% -6.7039%] (p = 0.00 < 0.05)
                        Performance has improved.

---

Here we can see up to 11% of performance improvement (for 100,000 pairs). The bad news here is that it takes up to 3.6 seconds to apply mutation set (without any other operations, like computations of mutation set). It's higher than current limit per block (65,536 accounts), but I think, it's too slow for fitting in 2 seconds per block limit.

**Comment by bobbinth** [comment]
> Here we can see up to 11% of performance improvement (for 100,000 pairs). The bad news here is that it takes up to 3.6 seconds to apply mutation set (without any other operations, like computations of mutation set). It's higher than current limit per block (65,536 accounts), but I think, it's too slow for fitting in 2 seconds per block limit.

To clarify: how big are the trees in these benchmarks and how many leaves are we updating? For now, we can assume that we probably won't be updating more than 10K leaves per block - so, I'd benchmark this number. Could you benchmark how long this takes on a fairly large tree (e.g., 1M or 10M entries)?

If this still takes a long time, we may need to start thinking about how to optimize the `apply_mutations()` method.

**Comment by polydez** [comment]
> To clarify: how big are the trees in these benchmarks and how many leaves are we updating? For now, we can assume that we probably won't be updating more than 10K leaves per block - so, I'd benchmark this number. Could you benchmark how long this takes on a fairly large tree (e.g., 1M or 10M entries)?

In this benchmarks we set up trees to have 10x more leaves than updates (10K leaves for 1K updates and 1M leaves for 100K updates). I will try separate benchmarks for 1M and 10M trees with 10K updating leaves.

**Comment by polydez** [comment]
> If this still takes a long time, we may need to start thinking about how to optimize the `apply_mutations()` method.

We should also benchmark `compute_mutations()` method and add metrics for all block building phases because `apply_mutations()` is just a part of block building.

**Comment by polydez** [comment]
@bobbinth, I have first numbers.

For tree with 1M leaves, benchmark 10K updates:

---
apply_mutations/SimpleSmt: apply_mutations/10000
                        time:   [2.1603 s 2.2503 s 2.3556 s]
---

**Comment by bobbinth** [comment]
> apply_mutations/SimpleSmt: apply_mutations/10000
> time: [2.1603 s 2.2503 s 2.3556 s]

This is pretty slow (probably around 10x slower than I thought it would be). This basically means we spend 0.1 ms per updated leaf. Could you check what's the bottleneck here?

**Comment by polydez** [comment]
> > apply_mutations/SimpleSmt: apply_mutations/10000
> > time: [2.1603 s 2.2503 s 2.3556 s]
> 
> This is pretty slow (probably around 10x slower than I thought it would be). This basically means we spend 0.1 ms per updated leaf. Could you check what's the bottleneck here?

The bottleneck here is `BTreeMap` itself. Applying mutations set is just inserting/updating values in `BTreeMap`. There is no duplicates because mutations set contains of two maps with unique keys.

I will try to rewrite `SimpleSmt` to hashmaps, as proposed in one of issues, and measure performance change.

**Comment by polydez** [comment]
I've rewritten `SimpleSmt` and `Smt` to hashbrown hashmaps, as well, as `MutationSet`. This dramatically improved performance:

---

apply_mutations/SimpleSmt: apply_mutations/10000
                        time:   [154.52 ms 163.09 ms 174.46 ms]
                        change: [-93.242% -92.753% -92.142%] (p = 0.00 < 0.05)
                        Performance has improved.

---

But overall benchmarking takes more than half of hour for 10 samples. So probably `compute_mutations` still takes the most of the time, but this time is not included into the benchmark. I will try to measure it.

**Comment by bobbinth** [comment]
> I've rewritten `SimpleSmt` and `Smt` to hashbrown hashmaps, as well, as `MutationSet`. This dramatically improved performance:

Very nice! Could you create a separate PR for this? Not sure if possible, but would be awesome to have this behind a feature flag.

Also, we are mostly interested in `Smt` (rather than `SimpleSmt`) as we'll be moving to use `Smt` for both accounts and nullifiers.

> So probably `compute_mutations` still takes the most of the time, but this time is not included into the benchmark. I will try to measure it.

We are working on parallelizing this - so, hopefully, we should be able to run `compute_mutations` on multiple cores.

**Comment by bobbinth** [comment]
> Do we rely on ordering of iterators over inner nodes or leaves? This is the only obstacle I can see now for full migration of `Smt`/`SimpleSmt` to hashmaps.

Hmmm - good question. I don't think we should be, but I'm not 100% sure we are not.

**Comment by bobbinth** [comment]
> > Do we rely on ordering of iterators over inner nodes or leaves? This is the only obstacle I can see now for full migration of `Smt`/`SimpleSmt` to hashmaps.
> 
> Hmmm - good question. I don't think we should be, but I'm not 100% sure we are not.

One thing that comes to mind is that order of leaves is important during `Smt` construction (specifically for the multi-threaded version) and so, the order in which we serialize leaves would also be important. So, maybe we do need to preserve the iteration order. But now that I think of it, I'm not sure the current iterators have the same order as what is assumed by the constructor.

**Comment by bobbinth** [comment]
btw, I think we already have some benchmarks for computing mutations [here](https://github.com/0xPolygonMiden/crypto/blob/next/src/main.rs).

**Comment by polydez** [comment]
Benchmarking of `compute_mutations()` method (tree size and mutation count are the same, as before: 1M and 10K):

---

compute_mutations/SimpleSmt: compute_mutations/10000
                        time:   [2.6533 s 2.6876 s 2.7379 s]

---


> btw, I think we already have some benchmarks for computing mutations [here](https://github.com/0xPolygonMiden/crypto/blob/next/src/main.rs).

Ah, didn't find this, thank you!

**Comment by polydez** [comment]
> One thing that comes to mind is that order of leaves is important during `Smt` construction (specifically for the multi-threaded version) and so, the order in which we serialize leaves would also be important.

If it's important, we can sort leaves before serialization.

File: src/merkle/smt/full/tests.rs [hunk 5]
```diff
@@ -412,21 +414,49 @@ fn test_prospective_insertion() {
 
     let mutations = smt.compute_mutations(vec![(key_1, value_1)]);
     assert_eq!(mutations.root(), root_1, "prospective root 1 did not match actual root 1");
-    smt.apply_mutations(mutations).unwrap();
+    let revert = smt.apply_mutations_with_reversion(mutations).unwrap();
```
Comment (bobbinth) [inline src/merkle/smt/full/tests.rs:417]: With this change (and other changes in this file) we are no longer testing the `apply_mutations()` method by itself. Since these are implemented as two different methods, we should probably test both? Or maybe just the newly added tests should be enough?

**Review by bobbinth — COMMENTED** [review]
Looks good! Thank you! I left a couple of small comments inline.

File: src/merkle/smt/full/tests.rs (hunk 5 above)
Comment (polydez) [reply src/merkle/smt/full/tests.rs:417]: Yes, it's better to tests both, thank you for noticing!

File: benches/smt-mutations.rs
[the remark this answers is not in the document]
Comment (polydez) [reply benches/smt-mutations.rs:1]: This is very strange, but I faced different results in benchmarks using criterion.rs and ad-hoc ones in main.rs:

criterion.rs: 2.0870 s
main.rs: ~0.2 s

Benchmarks using criterion.rs run 10x slower (it runs samples for 10 times, but this doesn't seem to be the reason for 10x slower performance). I spent whole day trying to improve our ad-hoc benchmarks to make them very similar to ones in benchmarks dir, but the results are still the same.

Will investigate more.

[the remark this answers is not in the document]
Comment (polydez) [reply benches/smt-mutations.rs:1]: I think, I understand, what happens with benchmarks. When I was running criterion benchmarks, they take about hour to execute. That's why I switched to another processes and benchmarks were running on background. But benchmarks in main.rs run faster and I usually was waiting for finishing keeping them on foreground. I think, OS scheduler might move background tasks to energy efficient CPU cores, but foreground to performant ones. This can explain, why results so different. I tried benchmarks in main.rs moved to background and got results, which are much slower than before without changes in SMT implementation:

---

Running a construction benchmark:
Constructed a SMT with 10000000 key-value pairs in 3396.505 seconds
Number of leaf nodes: 10000000

Running an insertion benchmark:
An average insertion time measured by 1000 inserts into an SMT with 10000000 leaves is 1165.570 microseconds

Running a batched insertion benchmark:
An average insert-batch computation time measured by a 10000-batch into an SMT with 10001000 leaves over 6173.059 milliseconds is 0.617 milliseconds
An average insert-batch application time measured by a 10000-batch into an SMT with 10001000 leaves over 1103.965 milliseconds is 0.110 milliseconds
An average batch insertion time measured by a 10k-batch into an SMT with 10001000 leaves totals to 7277.024 milliseconds

Running a batched update benchmark:
An average update-batch computation time measured by a 10000-batch into an SMT with 10011000 leaves over 5584.594 milliseconds is 0.558 milliseconds
An average update-batch application time measured by a 10000-batch into an SMT with 10011000 leaves over 708.746 milliseconds is 0.071 milliseconds
An average batch update time measured by a 10k-batch into an SMT with 10011000 leaves totals to 6293.341 milliseconds

Running a proof generation benchmark:
An average proving time measured by 20 value proofs in an SMT with 10008937 leaves in 10.300 microseconds

---

**Review by bobbinth — APPROVED** [review]
All looks good! Thank you!
